File: python/tennis_analyzer/phase_detector.py
"""Phase-detection heuristics: auto-locate the trophy and contact frames
from a per-frame pose time series (Phase 2).

Prototyped and unit-tested here first per the playbook's Python-before-Swift
workflow (docs/tennis-serve-app-full-playbook.md §3, §10) — ported 1:1 to
PhaseDetector.swift once validated. Vision itself has no Python equivalent
and is implemented directly in Swift; this module only covers the numeric
heuristic layered on top of its output.
"""

import math
from dataclasses import dataclass

from tennis_analyzer.geometry import angle_at_vertex

MIN_VALID_FRAMES = 3
DEFAULT_MIN_CONFIDENCE = 0.3


@dataclass
class PhaseDetectionResult:
    trophy_frame_index: int
    contact_frame_index: int


def speed(a, b, dt):
    """Finite-difference speed between two (x, y) points over dt seconds.

    Returns 0.0 if dt is zero or negative (degenerate).
    """
    if dt <= 0:
        return 0.0
    return math.hypot(b[0] - a[0], b[1] - a[1]) / dt


def hitting_side(left_wrist_speeds, right_wrist_speeds):
    """Which wrist has the higher peak speed across the sequence.

    Used only as a fallback when no handedness setting is available.
    Returns "left" or "right"; ties favor "right".
    """
    left_peak = max(left_wrist_speeds, default=0.0)
    right_peak = max(right_wrist_speeds, default=0.0)
    return "left" if left_peak > right_peak else "right"


def detect_contact_frame(wrist_speeds):
    """Index of peak wrist speed."""
    return max(range(len(wrist_speeds)), key=lambda i: wrist_speeds[i])


def detect_trophy_frame(knee_angles, before_index):
    """Index of the most-bent knee, restricted to frames strictly before
    `before_index` (the detected contact frame).
    """
    return min(range(before_index), key=lambda i: knee_angles[i])


def _is_valid(frame, joint_names, min_confidence):
    for name in joint_names:
        joint = frame.get(name)
        if joint is None or joint[2] < min_confidence:
            return False
    return True


def detect_phases(frames, hitting_side, frame_duration, min_confidence=DEFAULT_MIN_CONFIDENCE):
    """Auto-locate the trophy and contact frame indices.

    `frames` is a list of per-frame joint dicts: joint name -> (x, y,
    confidence), one entry per decoded video frame, in image-pixel space.
    `hitting_side` is "left" or "right" (from the user's handedness
    setting). `frame_duration` is the time between frames, in seconds.

    Returns None if there aren't enough confidently-tracked frames to make
    a reliable call — callers should fall back to manual marking.
    """
    wrist_joint = f"{hitting_side}_wrist"
    hip_joint = f"{hitting_side}_hip"
    knee_joint = f"{hitting_side}_knee"
    ankle_joint = f"{hitting_side}_ankle"
    required_joints = (wrist_joint, hip_joint, knee_joint, ankle_joint)

    valid_indices = [
        i for i, frame in enumerate(frames)
        if _is_valid(frame, required_joints, min_confidence)
    ]
    if len(valid_indices) < MIN_VALID_FRAMES:
        return None

    wrist_speeds = [-math.inf] * len(frames)
    for prev_i, curr_i in zip(valid_indices, valid_indices[1:]):
        a = frames[prev_i][wrist_joint][:2]
        b = frames[curr_i][wrist_joint][:2]
        dt = frame_duration * (curr_i - prev_i)
        wrist_speeds[curr_i] = speed(a, b, dt)

    if all(s == -math.inf for s in wrist_speeds):
        return None

    contact_index = detect_contact_frame(wrist_speeds)

    valid_before_contact = {i for i in valid_indices if i < contact_index}
    if not valid_before_contact:
        return None

    knee_angles = [math.inf] * len(frames)
    for i in valid_before_contact:
        knee_angles[i] = angle_at_vertex(
            frames[i][hip_joint][:2], frames[i][knee_joint][:2], frames[i][ankle_joint][:2]
        )

    trophy_index = detect_trophy_frame(knee_angles, contact_index)

    return PhaseDetectionResult(
        trophy_frame_index=trophy_index,
        contact_frame_index=contact_index,
    )

File: ios-app/TennisAnalyzer/TennisAnalyzer/PhaseDetector.swift
import CoreGraphics
import Foundation

/// A single tracked joint's position (image-pixel space) and Vision's
/// per-joint detection confidence (0...1).
struct TrackedJoint {
    let point: CGPoint
    let confidence: Double
}

/// One frame's worth of tracked joints, keyed by name (e.g. "right_wrist").
typealias PoseFrame = [String: TrackedJoint]

struct PhaseDetectionResult {
    let trophyFrameIndex: Int
    let contactFrameIndex: Int
}

/// Per-joint confidence stats across a pose sequence.
struct JointStats {
    let framesPresent: Int
    let framesConfident: Int
    let averageConfidence: Double
    let minConfidence: Double
    let maxConfidence: Double
}

/// Always-populated diagnostics for a `detectPhases` run, including a
/// specific human-readable `failureReason` when `result` is nil.
struct PhaseDetectionDiagnostics {
    let result: PhaseDetectionResult?
    let failureReason: String?
    let totalFrames: Int
    let validFrameCount: Int
    let jointStats: [String: JointStats]
}

/// Phase-detection heuristics: auto-locate the trophy and contact frames
/// from a per-frame pose time series (Phase 2).
///
/// Mirrors `python/tennis_analyzer/phase_detector.py`, which was prototyped
/// and unit-tested first per the playbook's Python-before-Swift workflow.
enum PhaseDetector {
    static let minValidFrames = 3
    static let defaultMinConfidence = 0.3

    /// Finite-difference speed between two points over `dt` seconds.
    /// Returns 0 if `dt` is zero or negative (degenerate).
    static func speed(_ a: CGPoint, _ b: CGPoint, dt: Double) -> Double {
        guard dt > 0 else { return 0.0 }
        return Double(hypot(b.x - a.x, b.y - a.y)) / dt
    }

    /// Which wrist has the higher peak speed across the sequence. Used
    /// only as a fallback when no handedness setting is available. Ties
    /// favor "right".
    static func hittingSide(leftWristSpeeds: [Double], rightWristSpeeds: [Double]) -> String {
        let leftPeak = leftWristSpeeds.max() ?? 0.0
        let rightPeak = rightWristSpeeds.max() ?? 0.0
        return leftPeak > rightPeak ? "left" : "right"
    }

    /// Index of peak wrist speed.
    static func detectContactFrame(_ wristSpeeds: [Double]) -> Int {
        wristSpeeds.indices.max(by: { wristSpeeds[$0] < wristSpeeds[$1] })!
    }

    /// Index of the most-bent knee, restricted to frames strictly before
    /// `beforeIndex` (the detected contact frame).
    static func detectTrophyFrame(_ kneeAngles: [Double], beforeIndex: Int) -> Int {
        (0..<beforeIndex).min(by: { kneeAngles[$0] < kneeAngles[$1] })!
    }

    private static func isValid(_ frame: PoseFrame, jointNames: [String], minConfidence: Double) -> Bool {
        for name in jointNames {
            guard let joint = frame[name], joint.confidence >= minConfidence else { return false }
        }
        return true
    }

    /// Per-joint confidence stats across every joint name seen in `frames`.
    ///
    /// Covers all joints present in the data, not just the hitting-side
    /// required ones — this is what surfaces a handedness-setting mismatch
    /// (the real wrist tracked confidently on the *other* side) or a joint
    /// that drops out during a specific phase of the motion.
    static func summarizeJoints(_ frames: [PoseFrame], minConfidence: Double = defaultMinConfidence) -> [String: JointStats] {
        var jointNames = Set<String>()
        for frame in frames { jointNames.formUnion(frame.keys) }

        var stats: [String: JointStats] = [:]
        for name in jointNames {
            let confidences = frames.compactMap { $0[name]?.confidence }
            guard !confidences.isEmpty else { continue }
            stats[name] = JointStats(
                framesPresent: confidences.count,
                framesConfident: confidences.filter { $0 >= minConfidence }.count,
                averageConfidence: confidences.reduce(0, +) / Double(confidences.count),
                minConfidence: confidences.min()!,
                maxConfidence: confidences.max()!
            )
        }
        return stats
    }

    /// Like `detectPhases`, but always returns a populated diagnostics
    /// record — including a specific human-readable `failureReason` at
    /// whichever gate rejected the sequence, instead of a bare nil.
    static func diagnosePhases(
        frames: [PoseFrame],
        hittingSide: String,
        frameDuration: Double,
        minConfidence: Double = defaultMinConfidence
    ) -> PhaseDetectionDiagnostics {
        let jointStats = summarizeJoints(frames, minConfidence: minConfidence)

        func diagnostics(
            result: PhaseDetectionResult? = nil,
            failureReason: String? = nil,
            validFrameCount: Int = 0
        ) -> PhaseDetectionDiagnostics {
            PhaseDetectionDiagnostics(
                result: result,
                failureReason: failureReason,
                totalFrames: frames.count,
                validFrameCount: validFrameCount,
                jointStats: jointStats
            )
        }

        let wristJoint = "\(hittingSide)_wrist"
        let hipJoint = "\(hittingSide)_hip"
        let kneeJoint = "\(hittingSide)_knee"
        let ankleJoint = "\(hittingSide)_ankle"
        let requiredJoints = [wristJoint, hipJoint, kneeJoint, ankleJoint]

        let validIndices = frames.indices.filter {
            isValid(frames[$0], jointNames: requiredJoints, minConfidence: minConfidence)
        }
        guard validIndices.count >= minValidFrames else {
            return diagnostics(
                failureReason: "only \(validIndices.count) of \(frames.count) frames had all required "
                    + "\(hittingSide)-side joints (\(requiredJoints.joined(separator: ", "))) tracked above "
                    + "\(minConfidence) confidence (need >= \(minValidFrames))",
                validFrameCount: validIndices.count
            )
        }

        var wristSpeeds = [Double](repeating: -.infinity, count: frames.count)
        for (prevIndex, currIndex) in zip(validIndices, validIndices.dropFirst()) {
            let a = frames[prevIndex][wristJoint]!.point
            let b = frames[currIndex][wristJoint]!.point
            let dt = frameDuration * Double(currIndex - prevIndex)
            wristSpeeds[currIndex] = speed(a, b, dt: dt)
        }
        guard wristSpeeds.contains(where: { $0 != -.infinity }) else {
            return diagnostics(
                failureReason: "could not compute a wrist speed for any frame (\(validIndices.count) "
                    + "valid frame(s) found, but none formed a consecutive pair to measure movement between)",
                validFrameCount: validIndices.count
            )
        }

        let contactIndex = detectContactFrame(wristSpeeds)

        let validBeforeContact = Set(validIndices.filter { $0 < contactIndex })
        guard !validBeforeContact.isEmpty else {
            return diagnostics(
                failureReason: "no confidently-tracked frame occurs before the detected contact frame "
                    + "(index \(contactIndex))",
                validFrameCount: validIndices.count
            )
        }

        var kneeAngles = [Double](repeating: .infinity, count: frames.count)
        for index in validBeforeContact {
            let hip = frames[index][hipJoint]!.point
            let knee = frames[index][kneeJoint]!.point
            let ankle = frames[index][ankleJoint]!.point
            kneeAngles[index] = GeometryEngine.angleAtVertex(hip, knee, ankle)
        }

        let trophyIndex = detectTrophyFrame(kneeAngles, beforeIndex: contactIndex)

        return diagnostics(
            result: PhaseDetectionResult(trophyFrameIndex: trophyIndex, contactFrameIndex: contactIndex),
            validFrameCount: validIndices.count
        )
    }

    /// Auto-locates the trophy and contact frame indices.
    ///
    /// `frames` is one entry per decoded video frame, in image-pixel
    /// space. `hittingSide` is "left" or "right" (from the user's
    /// handedness setting). `frameDuration` is the time between frames,
    /// in seconds.
    ///
    /// Returns nil if there aren't enough confidently-tracked frames to
    /// make a reliable call — callers should fall back to manual marking.
    /// Use `diagnosePhases` for the reason why.
    static func detectPhases(
        frames: [PoseFrame],
        hittingSide: String,
        frameDuration: Double,
        minConfidence: Double = defaultMinConfidence
    ) -> PhaseDetectionResult? {
        diagnosePhases(frames: frames, hittingSide: hittingSide, frameDuration: frameDuration, minConfidence: minConfidence).result
    }
}
